conbench/entities/run.py

```diff
@@ -101,57 +103,93 @@ def create(data):
         run.save()
         return run
 
-    def get_baseline_run(self):
+    def get_baseline_run(
+        self,
+        commit_limit: int = 20,
+        on_default_branch: bool = False,
+        case_id: str = None,
+        context_id: str = None,
+    ) -> Optional["Run"]:
+        """Return the closest ancestor of this Run, where the ancestor run:
+
+        - is in the last ``commit_limit`` commits of this Run's commit ancestry
+        - is on the default branch, if on_default_branch is True (else on any branch)
+        - shares this Run's hardware
+        - has a BenchmarkResult with the given case_id/context_id, if those are given
+        - if they aren't given, has a BenchmarkResult with the case_id/context_id of
+          *any* of this Run's BenchmarkResults
+
+        Returns None if there are no matches. This could be a false negative if
+        ``commit_limit`` is too low (though note that the query takes longer with a
+        higher ``commit_limit``).
+
+        If there are multiple matches, prefer a Run with the same reason as this Run,
+        and then find the latest commit, finally tiebreaking by latest Run.timestamp.
```

Comment:
Could you expand a bit more about the reasoning behind matching run reasons? I suspect you're right to do this, but I'm curious what you're imagining might happen here that this would be disambiguating?

Reply:
This was inspired by the `conbench/tests/api/test_runs.py:: test_closest_commit_different_machines_should_omit_test_runs` test. In that test, we set up this run history (newer runs first). Each run is on a different commit:
```
run ID | machine | run reason
-------+---------+-----------
4      | m1      | commit
3      | m2      | commit
2      | m1      | testing
1      | m1      | commit
```

And that test has always ensured that the baseline of run 4 is run 1, not runs 2 or 3. I know the intention of the test was to exclude "testing" run reasons, but it also told me that we care more about preferring the same run reason than preferring a newer commit. 

This also means that if you select a "nightly" run, you'll get a "nightly" run as the baseline, even if there's a "commit" run on a commit in between. I thought that was nice, clean behavior. If you think it will lead to more confusion though, we could change it.

Reply:
> This also means that if you select a "nightly" run, you'll get a "nightly" run as the baseline, even if there's a "commit" run on a commit in between. I thought that was nice, clean behavior. If you think it will lead to more confusion though, we could change it.

I think this is totally fine to do for now (and I believe there are actually even no cases where this would happen right now) — but I do suspect if the future we actually will want to open this up a bit and allow for nightlies that have the same benchmarks as other types of runs (like a commit run) to be allowed to be considered as part of the history. There are some cases where this is more important than others (e.g. having a nightly run that has more robust history data because it also includes commit-to-default branch is one good example where having more flexibility is good. On the other hand, looking at per-commit benchmarks to detect a regression would be less useful to allow interspersed different run reasons).

I don't think we should change this PR or adjust this now, but we might (need to) do so in the future

Reply:
>to be allowed to be considered as part of the history

To clarify, this "prefer run reason" behavior only affects what happens when you're finding a baseline run (like clicking on the "compare ->" arrow in the webapp, or hitting the compare API). We still include ALL run reasons in the blue line in the timeseries plots (aka the history api), and in the baseline distribution when calculating z-scores.